gensim/models/translation_matrix.py

```diff
@@ -0,0 +1,329 @@
+
+
+class TranslationMatrix(utils.SaveLoad):
+    """
+    Objects of this class realize the translation matrix which map the source language
+    to the target language.
+     The main methods are:
+
+    1. constructor,
+    2. the `train` method, which initialize everything needed to build a translation matrix
+    3. the `translate` method, which given new word and its vector representation,
+    we map it to the other language space by computing z = Wx, then return the
+    word whose representation is close to z.
+
+    the details use seen the notebook (translation_matrix.ipynb)
+
+    >>> transmat = TranslationMatrix(word_pair, source_lang_vec, target_lang_vec)
+    >>> transmat.train(word_pair)
+    >>> translated_word = transmat.translate(words, topn=3)
+
+    """
+    def __init__(self, word_pair, source_lang_vec, target_lang_vec, random_state=None):
+        """
+        Initialize the model from a list pair of `word_pair`. Each word_pair is tupe
+         with source language word and target language word.
+
+        Examples: [("one", "uno"), ("two", "due")]
+
+        Args:
+            `word_pair` (list): a list pair of words
+            `source_lang_vec` (KeyedVectors): a set of word vector of source language
+            `target_lang_vec` (KeyedVectors): a set of word vector of target language
+        """
+
+        if len(word_pair[0]) != 2:
+            raise ValueError("Each training data item must contain two different language words.")
+
+        self.source_word, self.target_word = zip(*word_pair)
+        self.source_lang_vec = source_lang_vec
+        self.target_lang_vec = target_lang_vec
+
+        self.random_state = utils.get_random_state(random_state)
+        self.translation_matrix = None
+        self.source_space = None
+        self.target_space = None
+
+    def train(self, word_pair):
+        """
+        build the translation matrix that mapping from source space to target space.
+
+        Args:
+            `word_pair` (list): a list pair of words
+            `source_space` (Space object): source language space
+            `target_space` (Space object): target language space
+
+        Returns:
+            `translation matrix` that mapping from the source language to target language
+        """
+        self.source_space = Space.build(self.source_lang_vec, set(self.source_word))
+        self.target_space = Space.build(self.target_lang_vec, set(self.target_word))
+
+        self.source_space.normalize()
+        self.target_space.normalize()
+
+        m1 = self.source_space.mat[[self.source_space.word2index[item] for item in self.source_word], :]
+        m2 = self.target_space.mat[[self.target_space.word2index[item] for item in self.target_word], :]
+
+        self.translation_matrix = np.linalg.lstsq(m1, m2, -1)[0]
+        return self.translation_matrix
+
+    def save(self, *args, **kwargs):
+        """
+        Save the model to file but ignoring the souce_space and target_space
```

Comment:
`source_space`

Reply:
Thanks, I will fix all the grammar errors and typos here and below.